### PR TITLE
Feature: custom config for coding and jam-codec compatibility

### PR DESCRIPTION
**Pull request by xDimon**
Added and implemented following options

```
option(JAM_COMPATIBLE "Build compatible with JAM-codec" OFF)
option(CUSTOM_CONFIG_SUPPORT "Support custom config of streams" OFF)
```

### CODE REVIEW

File: include/scale/detail/jam_compact_integer.hpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+   * Encodes any integer type to jam-compact-integer representation
+   * @tparam T integer type
+   * @tparam S output stream type
+   * @param value integer value
+   * @return byte array representation of value as jam-compact-integer
+   */
+  template <typename T, typename S>
+    requires(std::unsigned_integral<T> or std::is_same_v<T, CompactInteger>)
+  void encodeJamCompactInteger(T integer, S &s) {
+    size_t value;
```
Comment (turuslan) [inline include/scale/detail/jam_compact_integer.hpp:30]:
```suggestion
    uint64_t value;
```

Comment (xDimon) [reply include/scale/detail/jam_compact_integer.hpp:30]: Why? `size_t` represents biggest unsigned integer type considering of platform.

File: include/scale/configurable.hpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+   public:
+    Configurable() = default;
+    ~Configurable() = default;
+
+    template <typename T>
+    explicit Configurable(const T &config) : config_(std::cref(config)) {}
+
+#ifdef CUSTOM_CONFIG_ENABLED
+    template <typename T>
+    const T &getConfig() const {
```
Comment (turuslan) [inline include/scale/configurable.hpp:25]: Why allow inconsistent `T` in constructor and `getConfig`?

File: include/scale/configurable.hpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    void addConfig(const ConfigT &config) {
+      auto [_, added] =
+          configs_.emplace(typeid(ConfigT).hash_code(), std::cref(config));
+      if (not added) {
+        throw std::runtime_error(
+            "Stream can be configured by different custom config types only");
+      }
+    }
+
+    std::unordered_map<typeid_hash, std::any> configs_{};
```
Comment (turuslan) [inline include/scale/configurable.hpp:69]: https://en.cppreference.com/w/cpp/types/type_index
```suggestion
    std::unordered_map<std::type_index, std::any> configs_{};
```

**Review by turuslan — APPROVED** [review]

**Review by turuslan — APPROVED** [review]

File: include/scale/scale_encoder_stream.hpp [hunk 4]
```diff
@@ -305,4 +330,4 @@ namespace scale {
     return s << static_cast<std::underlying_type_t<E>>(v);
   }
 
-}  // namespace scale
+}  // namespace scale
```
Comment (kamilsa) [inline include/scale/scale_encoder_stream.hpp:333]: Empty line

**Review by kamilsa — DISMISSED** [review]

**Review by kamilsa — APPROVED** [review]

**Review by kamilsa — APPROVED** [review]

**Pull request merged**